torch/csrc/distributed/c10d/ProcessGroup.cpp

```diff
@@ -186,4 +208,15 @@ void ProcessGroup::init() {
   C10_LOG_API_USAGE_ONCE(fmt::format("c10d.process_group_{}", getBackendName()));
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroup::broadcast(
+    std::vector<at::Tensor>& tensors, const BroadcastOptions& opts) {
```

Comment:
So the python API here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L1162 does additional validation before calling the ProcessGroup.

Wouldn't this lead to some semantic differences where the same thing run in eager would have additional validation and might fail, whereas when you trace the op it doesn't have the additional validation and it might run into a weird error downstream instead?

Shouldn't we modify the python API to just directly call into this so that both eager and traced version have same behavior?

Reply:
This is a good question. During the first trace, the tracer will make sure that the inputs can pass those checks. During the replay, the tracer will make sure the subsequence inputs share critical properties as the first trace. Otherwise, it will start a new trace.